🚀 1단계 - 엔티티 매핑

hyeok7777 opened the pull request:
영한님 JPA 인강듣고 공부하고 정리하느라 좀 늦었습니다 🧐

hyeok7777 committed:
docs(README.md):base commit 추가

hyeok7777 committed:
feat(User.java):annotation 설정
test(UserTest.java):저장 및 수정 테스트

hyeok7777 committed:
feat(Question.java):annotation 설정
test(QuestionTest.java):저장 및 수정 테스트

hyeok7777 committed:
feat(Answer.java):annotation 설정
test(AnswerTest.java):저장 및 수정 테스트

hyeok7777 committed:
feat(BaseEntity.java):Base Entity 추가

hyeok7777 committed:
feat(DeleteHistory.java):annotation 추가

oper912 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,15 +1,22 @@
+    @Lob
     private String contents;
+    @Nullable
     private boolean deleted = false;
```
```suggestion
    @Column(name = "deleted", nullable = false)
    private boolean deleted;
```
ddl을 보면 deleted 컬럼을 `not null` 이네요 :)

oper912 commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,51 @@
+    public void setUpdatedAt(LocalDateTime updateAt) {
+        this.updateAt = updateAt;
+    }
+}
```
```suggestion
@EntityListeners(AuditingEntityListener.class)
@MappedSuperclass
public abstract class BaseEntity {

    @CreatedDate
    @Column(nullable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }
}
```
음 BaseEntity에 식별자를 두시는거보다는 Entity 마다 식별자 전략이 다를수도 있어서 개별적으로 추가해주는걸 추천드려요

미션의 요구사항의 DDL을 보면 createId, updateId는 없어도 될것같습니다
Entity Auditing을 위해서  `@EntityListeners(AuditingEntityListener.class)` 추가해주세요

oper912 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,9 +1,13 @@
 
-public class DeleteHistory {
+@Entity
+public class DeleteHistory extends BaseEntity {
```
1단계 요구사항 DDL을 확인해주시구 요구사항에 맞게 구현해주세요 :)

oper912 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,10 +1,21 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+    @Column(length = 100)
+    @Nullable
```
```suggestion
    @Column(length = 100, nullable = false)
```

oper912 commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,40 @@
+
     public static final Answer A1 = new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
     public static final Answer A2 = new Answer(UserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
+
```
별도로 AnswerRepositoryTest 를 생성해주세요

oper912 replied:
AnswerTest에서는 추후 미션에서 추가될 Answer에 존재하는 비즈니스로직을 테스트해주세요

oper912 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,15 +1,22 @@
 
-public class Answer {
+@Entity
+public class Answer extends BaseEntity {
```
Entity에서 setter는 모두 제거해주세요 
[Setter금지](https://velog.io/@aidenshin/%EB%82%B4%EA%B0%80-%EC%83%9D%EA%B0%81%ED%95%98%EB%8A%94-JPA-%EC%97%94%ED%8B%B0%ED%8B%B0-%EC%9E%91%EC%84%B1-%EC%9B%90%EC%B9%99)

[equals hashcode를 재정의](https://velog.io/@park2348190/JPA-Entity%EC%9D%98-equals%EC%99%80-hashCode)해주세요

hyeok7777 replied:
set deleted 같은 경우는 서비스로직쪽에서 쓰이고 있는 곳이 있어 일단 놔뒀습니다.

oper912 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,9 +1,13 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
     private ContentType contentType;
```
enum을 숫자가 아니라 문자로 관리할 수 있는 애노테이션을 활용해주세요

oper912 reviewed (changes requested):
안녕하세요 진혁님 :)
JPA 미션 함께 하게된 권순호라고 합니다 🙂

1단계 진행해주셨는데 현재는 요구사항을 충족하지 못하는 부분이 보이는거같아요
참고하실만한 코멘트 남겼는데 확인해주시구 다시 한번 리뷰 요청 부탁드립니다 🙏

hyeok7777 committed:
feat(Answer.java, Question.java):annotation 수정
feat(BaseTimeEntity.java):파일명 수정, createAt, updateAt 제외 식별자 삭제
feat(DeleteHistory.java):요구사항DDL 반영
test(AnswerRepositoryTest.java):AnswerRepository에 대한 테스트 생성
feat(Answer):setter 삭제, equals, hashcode 재정의

oper912 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -5,13 +5,18 @@
 
 @Entity
-public class DeleteHistory extends BaseEntity {
+public class DeleteHistory extends BaseTimeEntity {
```
요구사항대로라면 `DeleteHistory`는 `BaseTimeEntity`를 상속받으면 안될 것 같아요!

oper912 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,36 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
 public class QuestionTest {
```
AnswerTest, AnswerRepositoryTest 분리해주신 것처럼 다른 도메인들도 마찬가지로 분리해주세요 🙂

oper912 reviewed (approved):
안녕하세요 진혁님:) 
피드백 반영 잘해주셨네요 🙂

추가적으로 코멘트 드린게 있는데 다음 단계 진행하시면서 반영하셔도 좋을 것 같아서
이만 머지하도록 하겠습니다

다음 단계 진행 부탁드려요~ 💪
